### DISCUSSION TITLE
How do I compare date-period intervals and time-period intervals?

### DISCUSSION THREAD

**Discussion by iainelder** [duckdb/duckdb#7170] [Q&A] [answered]
DuckDB syntax suggests there is just one interval type.

```sql
INTERVAL n UNIT
```

where `n` is an integer and `UNIT` is a unit of time such as `SECOND`, `MINUTE`, `HOUR`, `DAY`, and so on.

Then in the [interval documentation](https://duckdb.org/docs/sql/data_types/interval) there is this phrase that suggests there are actually two types:

> Intervals can also be used to specify a time period rather than a date period

And DuckDB seems to behave as if time-period intervals and date-period intervals are two different types that can't be added and subtracted.

```sql
SELECT interval 1 day AS A, interval 24 hours AS B, A - B AS C, A + B AS D;
```

|   A   |    B     |        C        |       D        |
|-------|----------|-----------------|----------------|
| 1 day | 24:00:00 | 1 day -24:00:00 | 1 day 24:00:00 |

This behavior surprises me. I can't see what it would be used for.

How do I make it behave so that `A - B = 0` and `A + B = interval 2 days = interval 48 hours`?

**Comment by iainelder** [comment]
Here's another example that more closely represents the real data I'm working with.

I have two timestamps that represent one duration. I have a count of minutes that represents another duration. I want to find the difference in the length of each duration.

I try to do that with this query and I for the sample data I expect the answer to be 1 minute, because there are (normally) 1440 minutes in a day.

```sql
SELECT
  '2023-04-20T00:00:00'::TIMESTAMP AS StartDate,
  '2023-04-21T00:00:00'::TIMESTAMP AS EndDate,
  to_minutes(CAST(1439 AS BIGINT)) AS DurationA,
  EndDate - StartDate AS DurationB,
  DurationA - DurationB AS DurationDiff;
```

But instead I get something I don't understand.

|      StartDate      |       EndDate       | DurationA | DurationB |   DurationDiff   |
|---------------------|---------------------|-----------|-----------|------------------|
| 2023-04-20 00:00:00 | 2023-04-21 00:00:00 | 23:59:00  | 1 day     | -1 days 23:59:00 |

**Comment by Mytherin** [comment, marked as answer]
Interval is a single type that consists of three separate components: `month`, `day` and `microsecond`. The reason these are separate components is because there is no "correct" translation between these components. A month is not a fixed amount of days, and a day is not a fixed amount of microseconds. Having these three separate components enables you to correctly model date transformations with different units, e.g.:

```sql
select date '2000-01-01' + interval 1 months as m;
┌────────────┐
│     m      │
│    date    │
├────────────┤
│ 2000-02-01 │
└────────────┘
select date '2000-01-01' + interval 2 months as m;
┌────────────┐
│     m      │
│    date    │
├────────────┤
│ 2000-03-01 │
└────────────┘
```

The DuckDB interval type is equivalent to the [Postgres interval type](https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-interval/).

If you want to find the *difference* between two timestamps, then you likely are not looking for intervals. That is because the difference between two dates depends on the unit you are looking for.

For example, the difference between `2000-01-01` and `2000-02-01` is either one month, or 31 days, or 744 hours. The difference between `2000-02-01` and `2000-03-01` is one month, 29 days or 696 hours. 

You are probably looking for the `datediff` function, which allows you to specify the unit for the difference that you want to measure:

```sql
D select datediff('month', date '2000-02-01', date '2000-03-01') as diff;
┌───────┐
│ diff  │
│ int64 │
├───────┤
│     1 │
└───────┘
D select datediff('day', date '2000-02-01', date '2000-03-01') as diff;
┌───────┐
│ diff  │
│ int64 │
├───────┤
│    29 │
└───────┘
D select datediff('hour', date '2000-02-01', date '2000-03-01') as diff;
┌───────┐
│ diff  │
│ int64 │
├───────┤
│   696 │
└───────┘

```

**Comment by iainelder** [comment]
That's a great explanation. Thanks, @Mytherin .

Datediff is exactly the function I was looking for, just like in [SQL Server](https://learn.microsoft.com/en-us/sql/t-sql/functions/datediff-transact-sql?view=sql-server-ver16).

So to get the answer I expect in my second example, I would write this instead.

```sql
SELECT
  '2023-04-20T00:00:00'::TIMESTAMP AS StartDate,
  '2023-04-21T00:00:00'::TIMESTAMP AS EndDate,
  1439 AS DurationA,
  datediff('minute', StartDate, EndDate) AS DurationB,
  DurationB - DurationA AS DurationDiff;
```

|      StartDate      |       EndDate       | DurationA | DurationB | DurationDiff |
|---------------------|---------------------|-----------|-----------|--------------|
| 2023-04-20 00:00:00 | 2023-04-21 00:00:00 | 1439      | 1440      | 1            |
